// @payloadcms/plugin-form-builder/dist/utilities/lexical/converters/list.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ListHTMLConverter: function() {
        return ListHTMLConverter;
    },
    ListItemHTMLConverter: function() {
        return ListItemHTMLConverter;
    }
});
const _serializeLexical = require("../serializeLexical");
const ListHTMLConverter = {
    converter: async ({ converters, node, parent, submissionData })=>{
        const childrenText = await (0, _serializeLexical.convertLexicalNodesToHTML)({
            converters,
            lexicalNodes: node.children,
            parent: {
                ...node,
                parent
            },
            submissionData
        });
        return `<${node?.tag} class="${node?.listType}">${childrenText}</${node?.tag}>`;
    },
    nodeTypes: [
        'list'
    ]
};
const ListItemHTMLConverter = {
    converter: async ({ converters, node, parent })=>{
        const childrenText = await (0, _serializeLexical.convertLexicalNodesToHTML)({
            converters,
            lexicalNodes: node.children,
            parent: {
                ...node,
                parent
            }
        });
        if ('listType' in parent && parent?.listType === 'check') {
            return `<li aria-checked=${node.checked ? 'true' : 'false'} class="${'list-item-checkbox' + node.checked ? 'list-item-checkbox-checked' : 'list-item-checkbox-unchecked'}"
          role="checkbox"
          tabIndex=${-1}
          value=${node?.value}
      >
          {serializedChildren}
          </li>`;
        } else {
            return `<li value=${node?.value}>${childrenText}</li>`;
        }
    },
    nodeTypes: [
        'listitem'
    ]
};
