// @payloadcms/db-mongodb/dist/connect.js
/* eslint-disable @typescript-eslint/no-var-requires */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "connect", {
    enumerable: true,
    get: function() {
        return connect;
    }
});
const _mongoose = /*#__PURE__*/ _interop_require_default(require("mongoose"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const connect = async function connect(payload) {
    if (this.url === false) {
        return;
    }
    if (!payload.local && typeof this.url !== 'string') {
        throw new Error('Error: missing MongoDB connection URL.');
    }
    let urlToConnect = this.url;
    let successfulConnectionMessage = 'Connected to MongoDB server successfully!';
    const connectionOptions = {
        autoIndex: true,
        ...this.connectOptions,
        useFacet: undefined
    };
    if (process.env.NODE_ENV === 'test') {
        if (process.env.PAYLOAD_TEST_MONGO_URL) {
            urlToConnect = process.env.PAYLOAD_TEST_MONGO_URL;
        } else {
            connectionOptions.dbName = 'payloadmemory';
            const { MongoMemoryServer } = require('mongodb-memory-server');
            const getPort = require('get-port');
            const port = await getPort();
            this.mongoMemoryServer = await MongoMemoryServer.create({
                instance: {
                    dbName: 'payloadmemory',
                    port
                }
            });
            urlToConnect = this.mongoMemoryServer.getUri();
            successfulConnectionMessage = 'Connected to in-memory MongoDB server successfully!';
        }
    }
    try {
        this.connection = (await _mongoose.default.connect(urlToConnect, connectionOptions)).connection;
        const client = this.connection.getClient();
        if (!client.options.replicaSet) {
            this.transactionOptions = false;
            this.beginTransaction = undefined;
        }
        if (process.env.PAYLOAD_DROP_DATABASE === 'true') {
            this.payload.logger.info('---- DROPPING DATABASE ----');
            await _mongoose.default.connection.dropDatabase();
            this.payload.logger.info('---- DROPPED DATABASE ----');
        }
        this.payload.logger.info(successfulConnectionMessage);
    } catch (err) {
        this.payload.logger.error(`Error: cannot connect to MongoDB. Details: ${err.message}`, err);
        process.exit(1);
    }
};
