// @payloadcms/db-mongodb/dist/updateOne.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "updateOne", {
    enumerable: true,
    get: function() {
        return updateOne;
    }
});
const _handleError = /*#__PURE__*/ _interop_require_default(require("./utilities/handleError"));
const _sanitizeInternalFields = /*#__PURE__*/ _interop_require_default(require("./utilities/sanitizeInternalFields"));
const _withSession = require("./withSession");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const updateOne = async function updateOne({ id, collection, data, locale, req = {}, where: whereArg }) {
    const where = id ? {
        id: {
            equals: id
        }
    } : whereArg;
    const Model = this.collections[collection];
    const options = {
        ...(0, _withSession.withSession)(this, req.transactionID),
        lean: true,
        new: true
    };
    const query = await Model.buildQuery({
        locale,
        payload: this.payload,
        where
    });
    let result;
    try {
        result = await Model.findOneAndUpdate(query, data, options);
    } catch (error) {
        (0, _handleError.default)(error, req);
    }
    result = JSON.parse(JSON.stringify(result));
    result.id = result._id;
    result = (0, _sanitizeInternalFields.default)(result);
    return result;
};
