// @payloadcms/plugin-nested-docs/dist/fields/breadcrumbs.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    createBreadcrumbsField: function() {
        return createBreadcrumbsField;
    },
    default: function() {
        return _default;
    }
});
const createBreadcrumbsField = (relationTo, overrides = {})=>({
        name: 'breadcrumbs',
        localized: true,
        type: 'array',
        ...overrides || {},
        admin: {
            readOnly: true,
            ...overrides?.admin || {}
        },
        fields: [
            {
                name: 'doc',
                admin: {
                    disabled: true
                },
                maxDepth: 0,
                relationTo,
                type: 'relationship'
            },
            {
                fields: [
                    {
                        name: 'url',
                        admin: {
                            width: '50%'
                        },
                        label: 'URL',
                        type: 'text'
                    },
                    {
                        name: 'label',
                        admin: {
                            width: '50%'
                        },
                        type: 'text'
                    }
                ],
                type: 'row'
            },
            ...overrides?.fields || []
        ]
    });
const _default = createBreadcrumbsField;
