// @payloadcms/plugin-form-builder/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    fields: function() {
        return _fields.fields;
    },
    getPaymentTotal: function() {
        return _getPaymentTotal.getPaymentTotal;
    },
    default: function() {
        return _default;
    }
});
const _FormSubmissions = require("./collections/FormSubmissions");
const _Forms = require("./collections/Forms");
const _fields = require("./collections/Forms/fields");
const _getPaymentTotal = require("./utilities/getPaymentTotal");
const FormBuilder = (incomingFormConfig)=>(config)=>{
        const formConfig = {
            ...incomingFormConfig,
            fields: {
                checkbox: true,
                country: true,
                email: true,
                message: true,
                number: true,
                payment: false,
                select: true,
                state: true,
                text: true,
                textarea: true,
                ...incomingFormConfig.fields
            }
        };
        return {
            ...config,
            // admin: {
            //   ...config.admin,
            //   webpack: (webpackConfig) => ({
            //     ...webpackConfig,
            //     resolve: {
            //       ...webpackConfig.resolve,
            //       alias: {
            //         ...webpackConfig.resolve.alias,
            //         [path.resolve(__dirname, 'collections/FormSubmissions/hooks/sendEmail.ts')]: path.resolve(__dirname, 'mocks/serverModule.js'),
            //         [path.resolve(__dirname, 'collections/FormSubmissions/hooks/createCharge.ts')]: path.resolve(__dirname, 'mocks/serverModule.js'),
            //       },
            //     },
            //   })
            // },
            collections: [
                ...config?.collections || [],
                (0, _Forms.generateFormCollection)(formConfig),
                (0, _FormSubmissions.generateSubmissionCollection)(formConfig)
            ]
        };
    };
const _default = FormBuilder;
