// @payloadcms/plugin-nested-docs/dist/hooks/resaveChildren.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _populateBreadcrumbs = /*#__PURE__*/ _interop_require_default(require("../utilities/populateBreadcrumbs"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const resave = async ({ collection, doc, draft, pluginConfig, req })=>{
    const parentSlug = pluginConfig?.parentFieldSlug || 'parent';
    const parentDocIsPublished = doc._status === 'published';
    const children = await req.payload.find({
        collection: collection.slug,
        depth: 0,
        draft,
        locale: req.locale,
        req,
        where: {
            [parentSlug]: {
                equals: doc.id
            }
        }
    });
    const breadcrumbSlug = pluginConfig.breadcrumbsFieldSlug || 'breadcrumbs';
    try {
        await children.docs.reduce(async (priorSave, child)=>{
            await priorSave;
            const childIsPublished = typeof collection.versions === 'object' && collection.versions.drafts && child._status === 'published';
            if (!parentDocIsPublished && childIsPublished) return;
            await req.payload.update({
                id: child.id,
                collection: collection.slug,
                data: {
                    ...child,
                    [breadcrumbSlug]: await (0, _populateBreadcrumbs.default)(req, pluginConfig, collection, child)
                },
                depth: 0,
                draft: !childIsPublished,
                locale: req.locale,
                req
            });
        }, Promise.resolve());
    } catch (err) {
        req.payload.logger.error(`Nested Docs plugin has had an error while re-saving a child document${draft ? ' as draft' : ' as published'}.`);
        req.payload.logger.error(err);
    }
};
const resaveChildren = (pluginConfig, collection)=>async ({ doc, req })=>{
        await resave({
            collection,
            doc,
            draft: true,
            pluginConfig,
            req
        });
        if (doc._status === 'published') {
            await resave({
                collection,
                doc,
                draft: false,
                pluginConfig,
                req
            });
        }
        return undefined;
    };
const _default = resaveChildren;
