// @payloadcms/bundler-webpack/dist/configs/base.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getBaseConfig", {
    enumerable: true,
    get: function() {
        return getBaseConfig;
    }
});
const _findnodemodules = /*#__PURE__*/ _interop_require_default(require("find-node-modules"));
const _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
const _htmlwebpackplugin = /*#__PURE__*/ _interop_require_default(require("html-webpack-plugin"));
const _path = /*#__PURE__*/ _interop_require_default(require("path"));
const _webpack = /*#__PURE__*/ _interop_require_default(require("webpack"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const mockModulePath = _path.default.resolve(__dirname, '../mocks/emptyModule.js');
const mockDotENVPath = _path.default.resolve(__dirname, '../mocks/dotENV.js');
const mockDBAdapterPath = _path.default.resolve(__dirname, '../mocks/db-adapters.js');
const nodeModulesPaths = (0, _findnodemodules.default)({
    cwd: process.cwd(),
    relative: false
});
const getBaseConfig = (payloadConfig)=>{
    let nodeModulesPath = nodeModulesPaths.find((p)=>{
        const guess = _path.default.resolve(p, 'payload/dist/admin');
        if (_fs.default.existsSync(guess)) {
            return true;
        }
        return false;
    });
    if (!nodeModulesPath) {
        nodeModulesPath = process.cwd();
    }
    const adminFolderPath = _path.default.resolve(nodeModulesPath, 'payload/dist/admin');
    const config = {
        entry: {
            main: [
                adminFolderPath
            ]
        },
        module: {
            rules: [
                {
                    exclude: /\/node_modules\/(?!.+\.tsx?$).*$/,
                    test: /\.(t|j)sx?$/,
                    use: [
                        {
                            loader: require.resolve('swc-loader'),
                            options: {
                                jsc: {
                                    parser: {
                                        syntax: 'typescript',
                                        tsx: true
                                    }
                                }
                            }
                        }
                    ]
                },
                {
                    oneOf: [
                        {
                            test: /\.(?:ico|gif|png|jpg|jpeg|woff(2)?|eot|ttf|otf|svg)$/i,
                            type: 'asset/resource'
                        }
                    ]
                }
            ]
        },
        plugins: [
            new _webpack.default.ProvidePlugin({
                process: require.resolve('process/browser')
            }),
            new _webpack.default.DefinePlugin(Object.entries(process.env).reduce((values, [key, val])=>{
                if (key.indexOf('PAYLOAD_PUBLIC_') === 0) {
                    return {
                        ...values,
                        [`process.env.${key}`]: `'${val}'`
                    };
                }
                return values;
            }, {})),
            new _htmlwebpackplugin.default({
                filename: _path.default.normalize('./index.html'),
                template: payloadConfig.admin.indexHTML
            })
        ],
        resolve: {
            alias: {
                '@payloadcms/bundler-webpack': mockModulePath,
                '@payloadcms/db-mongodb': mockDBAdapterPath,
                '@payloadcms/db-postgres': mockDBAdapterPath,
                dotenv: mockDotENVPath,
                path: require.resolve('path-browserify'),
                payload$: mockModulePath,
                'payload-config': payloadConfig.paths.rawConfig,
                'payload-user-css': payloadConfig.admin.css
            },
            extensions: [
                '.ts',
                '.tsx',
                '.js',
                '.json'
            ],
            fallback: {
                crypto: false,
                http: false,
                https: false
            },
            modules: [
                'node_modules',
                nodeModulesPath,
                _path.default.resolve(__dirname, '../../node_modules')
            ]
        }
    };
    return config;
};
